### PR TITLE
BlueStore: Improve fragmentation score metric

### PR DESCRIPTION
**Pull request by aclamk**
This is combination of fix and improvement to fragmentation score.
1) The fix
This eradicates error in calculation of maximum reference fragmentation score.
Max fragmentation is denominator of % calculation.
Basically, we used to assume that 1 byte wide chunks are possible.
Now, we assume that min_alloc_size is the minimal chunk possible.
https://docs.google.com/spreadsheets/d/16pfs-xoV6rhR3nHXmja-5iQ3pZW6AlSz_aZqi7_ZOFI/edit#gid=0

2) The improvement
The improvement downplays splitting large free chunks into multiple large free chunks.
It is intended to make fragmentation not increase much for as long as dominating chunks are >16MB.
It is when small chunks become frequent that we want fragmentation to increase.

Partially (and maybe completely) fixes:
https://tracker.ceph.com/issues/58022

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by ifed01** [comment]
@aclamk - you might want to fix unittest_alloc:
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] Allocator/AllocTest.test_alloc_fragmentation/0, where GetParam() = "stupid"
[  FAILED  ] Allocator/AllocTest.test_alloc_fragmentation/1, where GetParam() = "bitmap"

**Comment by aclamk** [comment]
> @aclamk - you might want to fix unittest_alloc: [ FAILED ] 2 tests, listed below: [ FAILED ] Allocator/AllocTest.test_alloc_fragmentation/0, where GetParam() = "stupid" [ FAILED ] Allocator/AllocTest.test_alloc_fragmentation/1, where GetParam() = "bitmap"

I have a problem fixing this test. 
Obviously, I can replace numbers.
But I do not understand value in this test.
Would you like me to modify it, so it will test behaviours like:
- complete fragmentation gives score close to 100%
- fragmentation is monotonic (same free space, less chunks is better)
- no fragmentation is 0
- few large allocations give negligent fragmentation 
?

**Comment by ifed01** [comment]
> ou like me to modify it, so it will test beha

Well, IIRC this case was originally intended for get_fragmentation() and relevant numbers were more or less explainable. Later we extended it with get_fragmentation_score() just to 
a) learn the difference
b) have any numbers that allows to track unexpected modifications.

Certainly having more specific test cases for get_fragmentation_score() would be great

File: src/test/objectstore/Allocator_test.cc [hunk 1]
```diff
@@ -329,11 +327,84 @@ TEST_P(AllocTest, test_alloc_fragmentation)
...
+  alloc->init_add_free(0, capacity);
+  // half (in 16 chunks) is completely full,
+  // other half completely fragmented, expect really high fragmentation score
+  for (uint64_t chunk = 0; chunk < capacity; chunk += capacity / 16) {
+    alloc->init_rm_free(chunk + capacity / 32, capacity / 32);
+    for (uint64_t pos = 0; pos < capacity / 32; pos += alloc_unit * 3) {
+      alloc->init_rm_free(chunk + pos, alloc_unit);
+    }
+  }
+  EXPECT_LT(0.9, alloc->get_fragmentation_score()); // 50% < f
```
Comment (ifed01) [inline src/test/objectstore/Allocator_test.cc:385]: nit: comment should be: 90% < f

**Review by ifed01 — APPROVED** [review]

**Comment by ljflores** [comment]
Rados suite review: https://pulpito.ceph.com/?branch=wip-yuri3-testing-2023-04-25-1147

Failures, unrelated:
    1. https://tracker.ceph.com/issues/59049
    2. https://tracker.ceph.com/issues/59192
    3. https://tracker.ceph.com/issues/59335
    4. https://tracker.ceph.com/issues/59380
    5. https://tracker.ceph.com/issues/58946
    6. https://tracker.ceph.com/issues/50371
    7. https://tracker.ceph.com/issues/59057
    8. https://tracker.ceph.com/issues/53345

Details:
    1. WaitReplicas::react(const DigestUpdate&): Unexpected DigestUpdate event - Ceph - RADOS
    2. cls/test_cls_sdk.sh: Health check failed: 1 pool(s) do not have an application enabled (POOL_APP_NOT_ENABLED)
    3. Found coredumps on smithi related to sqlite3Found coredumps on smithi related to sqlite3 - Ceph - Cephsqlite
    4. rados/singleton-nomsgr: test failing from "Health check failed: 1 full osd(s) (OSD_FULL)" and "Health check failed: 1 filesystem is offline (MDS_ALL_DOWN)" - Ceph - RADOS
    5. cephadm: KeyError: 'osdspec_affinity' - Ceph - Mgr - Dashboard
    6. Segmentation fault (core dumped) ceph_test_rados_api_watch_notify_pp - Ceph - RADOS
    7. rados/test_envlibrados_for_rocksdb.sh: No rule to make target 'rocksdb_env_librados_test' on centos 8 - Ceph - RADOS
    8. Test failure: test_daemon_restart (tasks.cephadm_cases.test_cli.TestCephadmCLI) - Ceph - Orchestrator

**Pull request merged**